### DISCUSSION TITLE
Removing a flatten Watermark from PDF

### DISCUSSION THREAD

**Discussion by eddprogrammer** [pymupdf/PyMuPDF#2105] [Looking for help] [answered]
Hi, I have a large number of PDF files with flatten watermark that need to be removed. The watermark exists as Form XObject in the PDF file, under the Resources\XObject group, see screenshot below. 
![image](https://user-images.githubusercontent.com/119921381/205843280-8f9a4e52-0822-4867-9321-1bcd7bdeae00.png)

I tried following the solution for a similar issue at the link below, but couldn't get it to work. That solution use the clean_contents, get_contents and read_contents method to remove the watermark, but I wasn't able to get my watermark using same method. Maybe it is because my watermark is under Resources, not Contents? 
https://github.com/pymupdf/PyMuPDF/discussions/1855

Really appreciate the help!

**Comment by JorjMcKie** [comment]
> Maybe it is because my watermark is under Resources, not Contents?

No, has nothing to do with this. Every XObject also will be referenced in the contents.

You did not really mention what you have tried. It is also not clear which watermark mechanism your PDFs are using. There are a handful of different ways to generate the **_impression_** of a "watermark". So there is no single way to remove them.
And some are no removable at all.

**Reply by eddprogrammer** [reply]
Thanks for the quick response! Agree it might look like "watermark," but it is not the same watermark as I would get when I use Adobe add watermark function. Here is [sample page](https://drive.google.com/file/d/1Wn-OHt7B2QeeQmnFdXSHaQtDZZTgC0-9/view?usp=sharing) I marked up for testing. 

I see the part i want to remove in the PDF Debugger 
0 TL
q
  BT
    0.753 0.753 0.753 rg
    /T1_0 100 Tf
    0 Tc
    0 Tw
    0 Ts
    100 Tz
    0 Tr
    0.7071 0.7071 -0.7071 0.7071 83.8309 337.1946 Tm
    (iDockom) Tj
  ET
  /TouchUp_TextEdit MP
  BT
    0 0 0 rg
    /T1_0 16 Tf
    100 10 Td
    (This document was ) Tj
  ET
Q

And I have tried using your sample code

`
page.clean_contents()
xref = page.get_contents()[0]  # get xref of resulting /Contents object
cont = bytearray(page.read_contents())  # read the contents source as a (modifyable) bytearray
if cont.find(b"iDockom") > 0:  # hoping to confirm the part i want to remove it available
    print("content found")
` 

But I wasn't able to find the string in "cont" or to remove it.

Thanks again for you help!

**Comment by JorjMcKie** [comment, marked as answer]
Whatever the PDF debugger is and what it pretends to show:
It did not display that page's `/Contents`!
Why don't you look yourself?
Doing `print(page.read_contents().decode())` will show a lot of vector graphics (line art) and no such text.
In the end, this means you cannot remove this approach of "watermarking".

**Reply by eddprogrammer** [reply]
It is the PDFBox Debugger available at https://pdfbox.apache.org/download.html. 

The reason it didn't display the page's /Contents was because I didn't expand the /Contents node. The screenshot below has that node expanded and selected - the content in the right panel is exactly what I will get by print(page.read_contents().decode()). 
![image](https://user-images.githubusercontent.com/119921381/206035532-e75259c9-7fd7-4c24-8b36-c04d21ca139b.png)
But the content (the one that looks like watermark) I want to remove is under the \Resources\XObject\FM0

The line art you are seeing is probably the word "Not Provided" on the page, which is actually not text. 
![image](https://user-images.githubusercontent.com/119921381/206035823-f72638d7-874a-4359-a590-3d32a496e72a.png)

Thanks again for bearing with me on this!

**Reply by JorjMcKie** [reply]
Ah, ok. If you indeed have located the responsible object for the watermark, then its removal is easy-peasy: delete the line with command `b"/Fm0 Do"`.

```python
cont = bytearray(doc.xref_stream(xref))  # modifyable version
p0 = cont.find(b"/Fm0")
p1 = cont.find(b"Do", p0)
cont[p0:p1+2] = b""
doc.update_stream(xref, cont)
```

**Reply by eddprogrammer** [reply]
Thanks! That works perfectly on the sample document!
